gdal/ogr/ogrsf_frmts/ngw/gdalngwdataset.cpp

```diff
@@ -0,0 +1,911 @@
+        return OGRERR_FAILURE;
+    }
+
+    if( iLayer < 0 || iLayer >= nLayers )
+    {
+        CPLError( CE_Failure, CPLE_AppDefined,
+            "Layer %d not in legal range of 0 to %d.", iLayer, nLayers-1 );
+        return OGRERR_FAILURE;
+    }
+
+    OGRNGWLayer *poLayer = static_cast<OGRNGWLayer*>(papoLayers[iLayer]);
+
+    if( poLayer->GetResourceId() != "-1" )
+    {
+        // For layers from server we can check permissions.
+
+        // We can skip check permissions here as papoLayers[iLayer]->Delete() will
+        // return false if no delete permission available.
+        FetchPermissions();
+
+        if( !stPermissions.bResourceCanDelete )
+        {
+            CPLError(CE_Failure, CPLE_AppDefined, "Operation not permitted.");
+            return OGRERR_FAILURE;
+        }
+    }
+
+    if( poLayer->Delete() )
+    {
+        delete poLayer;
+        memmove( papoLayers + iLayer, papoLayers + iLayer + 1,
+                 sizeof(void *) * (nLayers - iLayer - 1) );
+        nLayers--;
+    }
+
+    return OGRERR_NONE;
+}
+
+/*
+ * FillMetadata()
+ */
+void OGRNGWDataset::FillMetadata( const CPLJSONObject &oRootObject )
+{
+    std::string osCreateDate = oRootObject.GetString("resource/creation_date");
+    if( !osCreateDate.empty() )
+    {
+        GDALDataset::SetMetadataItem( "creation_date", osCreateDate.c_str() );
+    }
+    osName = oRootObject.GetString("resource/display_name");
+    SetDescription( osName.c_str() );
+    GDALDataset::SetMetadataItem( "display_name", osName.c_str() );
+    std::string osDescription = oRootObject.GetString("resource/description");
+    if( !osDescription.empty() )
+    {
+        GDALDataset::SetMetadataItem( "description", osDescription.c_str() );
+    }
+    GDALDataset::SetMetadataItem( "id", osResourceId.c_str() );
+
+    std::vector<CPLJSONObject> items =
+        oRootObject.GetObj("resmeta/items").GetChildren();
+
+    for( const CPLJSONObject &item : items )
+    {
+        std::string osSuffix = NGWAPI::GetResmetaSuffix( item.GetType() );
+        GDALDataset::SetMetadataItem( (item.GetName() + osSuffix).c_str(),
+            item.ToString().c_str(), "NGW" );
+    }
+}
+
+/*
+ * FlushMetadata()
+ */
+bool OGRNGWDataset::FlushMetadata( char **papszMetadata )
+{
+    if( !bMetadataDerty )
+    {
+        return true;
+    }
+
+    bool bResult = NGWAPI::FlushMetadata(osUrl, osResourceId, papszMetadata,
+        GetHeaders());
+    if( bResult )
+    {
+        bMetadataDerty = false;
+    }
+
+    return bResult;
+}
+
+/*
+ * SetMetadata()
+ */
+CPLErr OGRNGWDataset::SetMetadata( char **papszMetadata, const char *pszDomain)
+{
+    FetchPermissions();
+    if( !stPermissions.bMetadataCanWrite )
+    {
+        CPLError(CE_Failure, CPLE_AppDefined, "Operation not permitted.");
+        return CE_Failure;
+    }
+
+    CPLErr eResult = GDALDataset::SetMetadata(papszMetadata, pszDomain);
+    if( eResult == CE_None && EQUAL(pszDomain, "NGW") )
```

Comment:
pszDomain might be nullptr

Reply:
Fix per 4579aeab927d70ddab6142340bf5550c6789b42e